### PR TITLE
FIX: Retry logic on GET Requests

### PR DESCRIPTION
**Pull request by rymarczy**
Retry logic on GET Requests was not working as  expected, based on logging from initial application run. 

New retry logic should be more robust and provide expected behavior.

### CODE REVIEW

File: src/dmap_import/util_api.py [hunk 1]
```diff
@@ -47,23 +47,31 @@ def download_from_url(url: str, local_path: str) -> Optional[str]:
...
                 for file_chunk in response.iter_content(chunk_size=None):
                     # chunk_size=None will write chunks in whatever size they are received
                     local_file.write(file_chunk)
-        break
+
+            response.close()
+            break
+
+        except Exception as exception:
+            if retry_count == max_retries:
```
Comment (paulswartz) [inline src/dmap_import/util_api.py:66]: suggestion: I think you can replace this with a `for: else:` block:

```python
for try in range(max_tries):
  try:
    # do the thing
    break
  except Exception as e:
    # handle exception
else:
  # no breaks means we never succeeded
```

File: src/dmap_import/util_api.py [hunk 2]
```diff
@@ -131,24 +139,29 @@ def get_api_results(url: str, db_manager: DatabaseManager) -> List[ApiResult]:
     # will log and throw if 200 status_code not recieved
     # or if "success" attribute of ApiResponse is not True
     max_retries = 3
-    for retry_count in range(max_retries):
+    for retry_count in range(max_retries + 1):
```
Comment (paulswartz) [inline src/dmap_import/util_api.py:142]: suggestion: same here: `for: else:`

File: src/dmap_import/util_api.py (hunk 1 above)
Comment (paulswartz) [reply src/dmap_import/util_api.py:66]: If you want to use the exception outside the loop, you'd need to set it as a variable beforehand:
```
last_failure = None
for ...:
  try:
  ...
  except Exception as e:
    last_failure = (response, e)
    time.sleep(15)
else:
  (response, exception) = last_failure
  ...
```

Comment (rymarczy) [reply src/dmap_import/util_api.py:66]: The still throws a pylint message on `last_failure`:

E0601: Using variable 'last_failure' before assignment (used-before-assignment)

Comment (rymarczy) [reply src/dmap_import/util_api.py:66]: This configuration throws mypy errors.

Either:
```
error: "None" object is not iterable  [misc]
            (last_response, last_exception) = last_failure
```
with `last_failure = None` 

OR 

```
src/dmap_import/util_api.py:66: error: Incompatible types in assignment (expression has type
"tuple[Response, Exception]", variable has type "tuple[None, None]")  [assignment]
                last_failure = (response, exception)
                               ^~~~~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:73: error: "None" has no attribute "status_code"  [attr-defined]
                staus_code=last_response.status_code,
                           ^~~~~~~~~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:74: error: "None" has no attribute "text"  [attr-defined]
                response=last_response.text,
                         ^~~~~~~~~~~~~~~~~~
src/dmap_import/util_api.py:76: error: Argument 1 to "log_failure" of "ProcessLogger" has
incompatible type "None"; expected "Exception"  [arg-type]
            download_log.log_failure(last_exception)
                                     ^~~~~~~~~~~~~~
src/dmap_import/util_api.py:77: error: Exception must be derived from BaseException  [misc]
            raise last_exception
```
with `last_failure = (None, None)`

**Review by paulswartz — APPROVED** [review]

File: src/dmap_import/util_api.py [hunk 3]
```diff
@@ -47,23 +47,31 @@ def download_from_url(url: str, local_path: str) -> Optional[str]:
...
             with open(local_path, "wb") as local_file:
                 for file_chunk in response.iter_content(chunk_size=None):
                     # chunk_size=None will write chunks in whatever size they are received
                     local_file.write(file_chunk)
-        break
+
+            response.close()
+            break
+
+        except Exception as exception:
```
Comment (mzappitello) [inline src/dmap_import/util_api.py:65]: is there a specific exception that we would expect that would trigger a retry? we could then catch, log, and raise all others.

File: src/dmap_import/util_api.py [hunk 4]
```diff
@@ -47,23 +47,33 @@ def download_from_url(url: str, local_path: str) -> Optional[str]:
...
+                time.sleep(15)
+
+    else:
+        download_log.add_metadata(
+            staus_code=response.status_code,
+            response=response.text,
+        )
+        exception = requests.HTTPError(response.text)
+        download_log.log_failure(exception)
+        raise exception
```
Comment (rymarczy) [inline src/dmap_import/util_api.py:76]: Create new exception that won't have API keys in clear text.

File: src/dmap_import/util_api.py (hunk 1 above)
Comment (rymarczy) [reply src/dmap_import/util_api.py]: Was able to bring this back with a new `exception` work flow, previous exceptions had the possibility of leaking API keys in clear text.

File: src/dmap_import/util_api.py (hunk 3 above)
Comment (rymarczy) [reply src/dmap_import/util_api.py]: I think we'll need to let this run for a little while a document what kind of Exceptions/Error codes we run into, and then we could modify the retry logic based on that additional information.

**Review by paulswartz — APPROVED** [review]

File: src/dmap_import/util_api.py (hunk 1 above)
Comment (mzappitello) [reply src/dmap_import/util_api.py]: its too bad that you can't do `for: elif:` blocks. 

i think if we wanted this to pass mypy you have to include an extra if block in side of the else block

```
last_failure = None
for ...:
    try:
        ... 
    except Exception as e:
        last_failure = (response, e)
        time.sleep(15)
else: 
    if last_failure:
        (response, exception) = last_failure
        ...
```

which is kind of clunky and i wouldn't be _that_ surprised if pylint yelled at you to use an `elif`.

i also think it reads ok as it is.

File: src/dmap_import/util_api.py [hunk 5]
```diff
@@ -47,23 +47,33 @@ def download_from_url(url: str, local_path: str) -> Optional[str]:
...
+            response = requests.get(url, stream=True, timeout=15)
+            response.raise_for_status()
 
             # download file
             with open(local_path, "wb") as local_file:
                 for file_chunk in response.iter_content(chunk_size=None):
                     # chunk_size=None will write chunks in whatever size they are received
                     local_file.write(file_chunk)
-        break
+            response.close()
```
Comment (mzappitello) [inline src/dmap_import/util_api.py:61]: do we need to close this response if we hit an exception in the file download?

Comment (rymarczy) [reply src/dmap_import/util_api.py:61]: In my testing it hasn't caused any issues,  we probably don't really need to call it at all based on the library documentation:

```
close()[[source]](https://requests.readthedocs.io/en/latest/_modules/requests/models/#Response.close)

    Releases the connection back to the pool. Once this method has been called the underlying raw object must not be accessed again.

    Note: Should not normally need to be called explicitly.
```

Comment (mzappitello) [reply src/dmap_import/util_api.py:61]: looks like at worst we'd have some connection inefficiencies if something goes wrong in the reading. in normal operation, the connection will be released after the `iter_content` for loop (when all the data is consumed).

[src](https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow)

**Review by mzappitello — APPROVED** [review]

**Pull request merged**